### PR TITLE
Feature/stacking

### LINKED ISSUE #18: Basic Tool Stacking
To enhance what LD_PRELOAD offers, would be nice if we could have tools 1 and 2 both use GOTCHA, wrap the same symbols, and have them be stacked

For 0.1 we'll want this to work in the following way: if Tool1 wraps foo and then Tool2 wraps foo, Tool2 will see Tool1's foo as the "real" foo. This will lead to Tool2 calls Tool1 calls real_foo.

In the future, we'll want an interface for tool developers to say "I should happen before/after" other tools/tool categories.

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Fixes #18

Mainly committing so we have a shared test. This gives us the basic stacking functionality where if tool1 wraps before tool2, tool2 will call tool1's binds

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
@mplegendre : I made develop our development branch (odd, I know), had to close the old version of the PR and open this. Same contents

**Comment by DavidPoliakoff** [comment]
Code is currently polluted, I wouldn't merge this, but it _does_ now work with dlopen/dlsym. Only the dlsym is really tested at all, though. More to come soon, but we do have progress on dlsym, and maybe (?) progress on dlopen.

**Comment by DavidPoliakoff** [comment]
Bizarre, the tests are failing on Travis, but passing on my machine. Will dig in.

File: src/gotcha_utils.c [hunk 1]
```diff
@@ -57,3 +59,23 @@ void debug_init()
 
    debug_printf(0, "Gotcha debug initialized at level %d\n", debug_level);
 }
+
+void handle_libdl(){
```
Comment (mplegendre) [inline src/gotcha_utils.c]: Perhaps move to gotcha_dl.c with other dlopen handling code?

File: src/libc_wrappers.h [hunk 2]
```diff
@@ -58,7 +58,7 @@ void *gotcha_realloc(void* buffer, size_t size);
 void gotcha_free(void* free_me);
 void gotcha_memcpy(void* dest, void* src, size_t size);
 int gotcha_strncmp(const char* in_one, const char* in_two, int max_length);
-char *gotcha_strstr(char* searchIn, char* searchFor);
+const char *gotcha_strstr(const char* searchIn,const char* searchFor);
```
Comment (mplegendre) [inline src/libc_wrappers.h]: libc doesn't make the return value of strstr const.

File: test/dlopen/test_dlopen.c [hunk 3]
```diff
@@ -50,12 +58,8 @@ int main(int argc, char *argv[])
    int result;
 
    result = gotcha_wrap(funcs, 2, "dlopen_test");
-   if (result != 0) {
```
Comment (mplegendre) [inline test/dlopen/test_dlopen.c]: Why not check for error in the test?

Comment (DavidPoliakoff) [inline test/dlopen/test_dlopen.c]: Because, one of those functions isn't found, and shouldn't be (I believe, let me retest this). This is changed in another update I'm working on (the one with Wall Wextra weirdness).

Version there (will be merged soon into this one)

```
if(result != GOTCHA_FUNCTION_NOT_FOUND){
     fprintf(stderr, "GOTCHA should have failed to find a function, but found it\n");
     return -1;
   }
```

Comment (DavidPoliakoff) [inline test/dlopen/test_dlopen.c]: specifically, return_four doesn't exist

File: src/gotcha.c [hunk 4]
```diff
@@ -161,7 +178,9 @@ GOTCHA_EXPORT enum gotcha_error_t gotcha_wrap(struct gotcha_binding_t* user_bind
 
   for (lib_iter = _r_debug.r_map; lib_iter != 0; lib_iter = lib_iter->l_next) {
     debug_printf(2, "Looking for wrapped callsites in %s\n", LIB_NAME(lib_iter));
-    FOR_EACH_PLTREL(lib_iter, gotcha_wrap_impl, lib_iter, bindings, num_actions);
+    if(libraryFilterFunc(lib_iter)){
```
Comment (DavidPoliakoff) [inline src/gotcha.c]: Hmmm, I've got it as "filterFunc takes in a library, we only do our wrapping if the library meets the criteria that library defines." I must be missing something.

File: src/libc_wrappers.h (hunk 2 above)
Comment (DavidPoliakoff) [inline src/libc_wrappers.h]: Hmmm, I'll have to update the implementation, it discards constness (searchIn+i is a const char*, we return it as a char*)

Comment (DavidPoliakoff) [inline src/libc_wrappers.h]: actually, easy fix, the cast just has to be explicit. Fixed

Comment (DavidPoliakoff) [inline src/libc_wrappers.h]: Also that's a really gross behavior, but at least we do it accurately

File: src/gotcha_dl.c [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  filterLibrariesByName("libdl");
+  struct binding_t* tool_iter = get_bindings();
+  for(;tool_iter!=NULL;tool_iter = tool_iter->next_binding){
+    /**
+     * If you remove this comparison, you will overwrite
+     * libdl's dlsym relocation with gotcha's, leading to
+     * infinite recursion. This is cool, but inefficient.
+     */
+    if(gotcha_strcmp(tool_iter->tool->tool_name, "gotcha")!=0){
+      gotcha_wrap(tool_iter->user_binding,tool_iter->user_binding_size,tool_iter->tool->tool_name);
```
Comment (DavidPoliakoff) [inline src/gotcha_dl.c]: Re: C

```
#1  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#2  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#3  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#4  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#5  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#6  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#7  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#8  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#9  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#10 0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#11 0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
```

That goes on for at least 2000 stack frames. I'm honestly not quite sure why any more, give me a bit to reason about why this is what it is

Comment (DavidPoliakoff) [inline src/gotcha_dl.c]: Wait a tic, I just learned to read again, there's something desperately wrong in this code. Momento, wow that's broken. Good catch

File: src/gotcha.c (hunk 4 above)
Comment (mplegendre) [inline src/gotcha.c]: I assume the desired semantics are that someone could write a set of MPI wrappings, then use the filter function to limit the search to libmpi.so.

But this is the code that looks through callsites in every library for calls to libmpi.so.  If we filter here then we'll have filtered-callsites going to libmpi.so and unfiltered-callsites going to the tool.  I'm not imagining any cases where that's what the tool writer wants.

The libraryFilter could safely limit our gotcha_prepare_symbols step (not that that's a huge performance problem area).  Or it could filter the the callsites by only looking at libraries that depend on libmpi.so.  But I don't think it should be filtering all libraries.

File: src/gotcha_dl.c (hunk 5 above)
Comment (DavidPoliakoff) [inline src/gotcha_dl.c]: Hmmmm, but when I remove it, the code no longer works. Sorry for the spam, give me a bit to figure this out, this code is a little stale

File: src/gotcha.c (hunk 4 above)
Comment (DavidPoliakoff) [inline src/gotcha.c]: Oooof, I was not thinking straight when I wrote this code, you're right on that. Got callsite/callee mixed up again. Thanks for the catch, though I'm thrashing enough on this topic that I'll look a bit before acting on the feedback

**Comment by DavidPoliakoff** [comment]
Hey @mplegendre , I'll merge my current version of this, which will obsolete the review, but most of what you address is still relevant. It does have the fix for test_dlopen, though, and builds with Wall Werror Wextra. This way we're at least going off the same codebases. Feel free to keep the conversation going in the old threads though, they're equally applicable

File: src/gotcha.c [hunk 6]
```diff
@@ -149,18 +149,6 @@ GOTCHA_EXPORT enum gotcha_error_t gotcha_wrap(struct gotcha_binding_t* user_bind
       //printf("Writing %p, %zu bytes, page size %d remainder is %lx\n",prot_address,protect_size, page_size,((size_t)BOUNDARY_BEFORE(got,page_size)%page_size));
       int res = gotcha_mprotect((void*)prot_address,protect_size,PROT_READ | PROT_WRITE | PROT_EXEC );
       if(!res){
```
Comment (mplegendre) [inline src/gotcha.c]: Error test is backwards.  mprotect returns 0 on success, -1 on error.

Comment (DavidPoliakoff) [inline src/gotcha.c]: This and other genius software writing moves on "David writes C." I'll also clean up the commented printf

**Comment by DavidPoliakoff** [comment]
@jmellorcrummey  @mwkrentel

This is currently building on the PPC machines and passing unit tests (including your own) except in one corner case. I'd recommend holding off on trying anything today, we're looking to nail down some architectural things before exposing things to users, but we'll try to get this ready for friendly users ASAP.

**Comment by mplegendre** [comment]
@DavidPoliakoff, I'm losing track of where we are on this.  Can you give me an explicit ping if/when you want me to review this PR?  

Also, this seems to keep growing into a bit of a monster PR.  It may be worth closing this one and splitting it up into multiple PRs (stacking/ppc64le/tests/...) and re-submitting.  But even if we don't do that, let's not add new features (e.g., the discussed UI change) to this PR.

**Comment by DavidPoliakoff** [comment]
If you're game, I'll add do the (make_reverse_iterator->doubly linked list) change and then call this one ready. It is a monster PR, but if you think that the content is valid we'll close it and move on with more sensible ones going forward.

File: src/gotcha_dl.c [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+#include "tool.h"
+#include "libc_wrappers.h"
+#include "elf_ops.h"
+#include <dlfcn.h>
+
+void* _dl_sym(void* handle, const char* name, void* where);
+
+static void*(*orig_dlopen)(const char* filename, int flags);
+static void*(*orig_dlsym)(void* handle, const char* name);
+struct rev_iter* get_reverse_tool_iterator(struct binding_t* in){
```
[the remark this answers is not in the document]
Comment (DavidPoliakoff) [reply src/gotcha_dl.c]: Note for future review: we decided to merge this as is. It may be incorrect, but will be removed in the future interface refactoring
